larray/inout/hdf.py

```diff
@@ -73,53 +98,51 @@ def read_hdf(filepath_or_buffer, key, fill_value=nan, na=nan, sort_rows=False, s
-                if name is None:
-                    name = str(pd_obj.name)
-                if name == 'None':
-                    name = None
-                key = pd_obj.values
-                if 'dtype_kind' in attrs and attrs['dtype_kind'] == 'U':
-                    key = np.char.decode(key, 'utf-8')
-                axis = read_hdf(filepath_or_buffer, attrs['axis_key'])
-                res = LGroup(key=key, name=name, axis=axis)
-            elif _type == 'scalar':
-                res = pd_obj.values
-                # XXX : assert len(res) == 1 ?
-                if len(res) == 1:
-                    res = res[0]
         except KeyError:
             filepath = filepath_or_buffer if isinstance(filepath_or_buffer, HDFStore) else store.filename
             raise KeyError('No item with name {} has been found in file {}'.format(key, filepath))
+        attrs = store.get_storer(key).attrs
+        writer = attrs.writer if 'writer' in attrs else None
+        _type = _get_type_from_attrs(attrs)
+        _meta = attrs.metadata if 'metadata' in attrs else None
+        if _type == 'Array':
+            # cartesian product is not necessary if the array was written by LArray
+            cartesian_prod = writer != 'LArray'
+            res = df_asarray(pd_obj, sort_rows=sort_rows, sort_columns=sort_columns, fill_value=fill_value,
+                             parse_header=False, cartesian_prod=cartesian_prod)
+            if _meta is not None:
+                res.meta = _meta
+        elif _type == 'Axis':
+            if name is None:
+                name = str(pd_obj.name)
+            if name == 'None':
+                name = None
+            labels = pd_obj.values
+            if 'dtype_kind' in attrs and attrs['dtype_kind'] == 'U':
+                # this check is there because there are cases where dtype_kind is 'U' but pandas returns
+                # an array with object dtype containing bytes instead of a string array, and in that case
+                # np.char.decode does not work
+                # this is at least the case for Python2 + Pandas 0.24.2 combination
+                if labels.dtype.kind == 'O':
+                    labels = np.array([l.decode('utf-8') for l in labels], dtype='U')
+                else:
+                    labels = np.char.decode(labels, 'utf-8')
+            res = Axis(labels=labels, name=name)
+            res._iswildcard = attrs['wildcard']
+        elif _type == 'Group':
+            if name is None:
+                name = str(pd_obj.name)
+            if name == 'None':
+                name = None
+            key = pd_obj.values
+            if 'dtype_kind' in attrs and attrs['dtype_kind'] == 'U':
+                key = np.char.decode(key, 'utf-8')
+            axis = read_hdf(filepath_or_buffer, attrs['axis_key'])
+            res = LGroup(key=key, name=name, axis=axis)
+        elif _type in {cls.__name__ for cls in _supported_scalars_types}:
+            res = pd_obj.values
```

Comment:
Now it is there, let's finish this. I don't know when I will go back to work on PR #761 (partly because I don't like to work with pytables and prefer h5py but my intuition tells me we need pytables for our case).

Reply:
I thought h5py had the same features than pytables with a different API, except for querying (and optionally indexing) tables. And since we wouldn't need either feature, I would think using h5py would not be a problem. That said I never used h5py so I don't know for sure.

In either case, I wouldn't use h5py anyway because pandas relies on pytables and since we will have to keep the current code (using pytables indirectly) for backward compat for a while, depending on both pytables and h5py would be weird.